explorer/interpreter/type_checker.cpp

```diff
@@ -1094,31 +1095,44 @@ auto TypeChecker::ArgumentDeduction::Finish(TypeChecker& type_checker,
   return std::move(bindings);
 }
 
-namespace {
-// A queue implemented as a small-size-optimized fixed-size ring buffer.
-template <typename T>
-class SmallRingBufferQueue {
- public:
-  SmallRingBufferQueue(size_t max_size) : storage(max_size) {}
-  void Push(T v) {
-    CARBON_CHECK(size < static_cast<int>(storage.size()));
-    storage[(begin + size) % storage.size()] = std::move(v);
-    ++size;
-  }
-  auto Pop() -> std::optional<T> {
-    if (!size) {
-      return std::nullopt;
+// Look for a rewrite to use when naming the given interface member in a type
+// that has the given list of rewrites.
+static auto LookupRewrite(llvm::ArrayRef<RewriteConstraint> rewrites,
+                          Nonnull<const InterfaceType*> interface,
+                          Nonnull<const Declaration*> member)
+    -> std::optional<const RewriteConstraint*> {
+  if (!isa<AssociatedConstantDeclaration>(member)) {
+    return std::nullopt;
+  }
+
+  for (auto& rewrite : rewrites) {
+    if (ValueEqual(interface, &rewrite.constant->interface(), std::nullopt) &&
+        member == &rewrite.constant->constant()) {
+      // A ConstraintType can only have one rewrite per (interface, member)
+      // pair, so we don't need to check the rest.
+      return &rewrite;
     }
-    --size;
-    return std::move(storage[(begin + size) % storage.size()]);
   }
 
- protected:
-  llvm::SmallVector<T> storage;
-  int begin = 0;
-  int size = 0;
-};
-}  // namespace
+  return std::nullopt;
+}
+
+// Look for a rewrite to use when naming the given interface member in a type
+// declared with the given type-of-type.
+static auto LookupRewrite(Nonnull<const Value*> type_of_type,
+                          Nonnull<const InterfaceType*> interface,
+                          Nonnull<const Declaration*> member)
+    -> std::optional<const RewriteConstraint*> {
+  // Find the set of rewrites. Only ConstraintTypes have rewrites.
+  // TODO: If we can ever see an InterfaceType here, we should convert it to a
```

Comment:
Would it be worth a `CARBON_CHECK(!isa<InterfaceType>` to safeguard?

Reply:
Looks like this actually does happen. But, I'm just moving this function earlier in the file in this PR so I'll look into that separately.